implementations/rust/ockam/ockam_vault/src/traits/security_module.rs

```diff
@@ -0,0 +1,43 @@
+use crate::{PublicKey, SecretAttributes, Signature};
+use ockam_core::compat::boxed::Box;
+use ockam_core::{async_trait, KeyId, Result};
+
+/// A SecurityModule provides several functions related to secrets:
+///   - create and persist secrets
+///   - delete secrets
+///   - return the public key for a given key id
+///   - return the key id for a given public key
+///   - use a secret to sign a message
+///   - use a public key to verify a message signature
+///
+/// The concrete implementations for a security module can range from full KMSes like the AWS KMS
+/// to yubikeys or hardware security modules
+///
+#[async_trait]
+pub trait SecurityModule: Sync + Send {
+    /// Create a new secret and return its key id
+    async fn create_secret(&self, attributes: SecretAttributes) -> Result<KeyId>;
+
+    /// Get the public key from a secret
+    async fn get_public_key(&self, key_id: &KeyId) -> Result<PublicKey>;
+
+    /// Return the key id corresponding to a given public key
+    async fn get_key_id(&self, public_key: &PublicKey) -> Result<KeyId>;
```

Comment:
`SecretsStoreReader` also has it. Could you please explain why?

Reply:
That's because `SecretsStoreReader` is the high-level interface while the `SecurityModule` is part of the implementation, not exposed to the users of the `Vault`. And one delegates to the other.